### PR TITLE
Add ADC multisampling

### PR DESCRIPTION
**Pull request by Mat931**
# What does this implement/fix?

Added a new feature to sample the ADC multiple times in one reading and return the average. This is [recommended by Espressif](https://docs.espressif.com/projects/esp-idf/en/release-v4.4/esp32/api-reference/peripherals/adc.html#minimizing-noise) for noise reduction. Compared to increasing the update rate of the sensor this has the advantage that some math operations only need to be executed once and the averaging is done using integers.
For now it only takes multiple samples on the ESP32 if attenuation is not set to auto.
Tested on the ESP32, ESP8266 and RP2040. Testers needed for LibreTiny.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3664

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [x] ESP8266
- [x] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
external_components:
  - source: github://pr#6330
    components: [adc]
    refresh: always

sensor:
  - platform: adc
    id: my_voltage
    pin: GPIO36
    attenuation: 11db
    update_interval: 1s
    samples: 100
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/adc/adc_sensor.cpp [hunk 1]
```diff
@@ -165,53 +175,57 @@ float ADCSensor::sample() {
...
+        adc2_get_raw(this->channel2_, ADC_WIDTH_MAX_SOC_BITS, &raw);
+      }
+      if (raw == -1) {
+        return NAN;
+      }
+      sum += raw;
     }
-    if (output_raw_) {
-      return raw;
+    sum = sum / this->sample_count_;  // NOLINT(clang-analyzer-core.DivideZero)
```
Comment (X-Ryl669) [inline esphome/components/adc/adc_sensor.cpp]: There is a possible loss of precision here. I think it should read `sum = (sum + this->sample_count_ >> 1) / this->sample_count_` . For example, imagine you're reading [1, 4, 3] from the ADC. The obvious average is 2.66... But the sum will be 8/3 = 2 (with integer division) while it should have been 3 with the right rounding.

File: esphome/components/adc/adc_sensor.cpp [hunk 2]
```diff
@@ -246,8 +260,11 @@ float ADCSensor::sample() {
...
     delay(1);
     adc_select_input(4);
-
-    int32_t raw = adc_read();
+    uint32_t raw = 0;
+    for (uint8_t sample = 0; sample < this->sample_count_; sample++) {
+      raw += adc_read();
+    }
+    raw = raw / this->sample_count_;  // NOLINT(clang-analyzer-core.DivideZero)
     adc_set_temp_sensor_enabled(false);
```
Comment (X-Ryl669) [inline esphome/components/adc/adc_sensor.cpp:268]: Ditto

File: esphome/components/adc/adc_sensor.cpp [hunk 3]
```diff
@@ -268,15 +285,19 @@ float ADCSensor::sample() {
     adc_gpio_init(pin);
     adc_select_input(pin - 26);
 
-    int32_t raw = adc_read();
+    uint32_t raw = 0;
+    for (uint8_t sample = 0; sample < this->sample_count_; sample++) {
+      raw += adc_read();
+    }
+    raw = raw / this->sample_count_;  // NOLINT(clang-analyzer-core.DivideZero)
```
Comment (X-Ryl669) [inline esphome/components/adc/adc_sensor.cpp:293]: Ditto

File: esphome/components/adc/adc_sensor.cpp [hunk 4]
```diff
@@ -287,10 +308,17 @@ float ADCSensor::sample() {
...
 float ADCSensor::sample() {
-  if (output_raw_) {
-    return analogRead(this->pin_->get_pin());  // NOLINT
+  uint32_t raw = 0;
+  if (this->output_raw_) {
+    for (uint8_t sample = 0; sample < this->sample_count_; sample++) {
+      raw += analogRead(this->pin_->get_pin());  // NOLINT
+    }
+    return raw / this->sample_count_;  // NOLINT(clang-analyzer-core.DivideZero)
+  }
```
Comment (X-Ryl669) [inline esphome/components/adc/adc_sensor.cpp:318]: Here too (and next lines)

File: esphome/components/adc/adc_sensor.cpp (hunk 1 above)
Comment (Mat931) [reply esphome/components/adc/adc_sensor.cpp]: Thanks, fixed.

**Comment by edwardtfn** [comment]
You might wanna know we are playing with this solution on Blackymas/NSPanel_HA_Blueprint#1918 with positive results so far.
Just a couple of people testing, but it looks to be improving the temperature measurements in an Sonoff NSPanel (ESP32 with ESP-IDF).
I will try to get more people testing in the next couple of days.
Thanks for your work on this. 😃

**Comment by andythomas** [comment]
I would like to second this PR and share the data I collected using it with the NSPanel as already indicated in the last comment. A measurement without `samples` or any other filtering leads to the following plot. The red curve depicts the Gaussian fit to the data with a width of 0.2°C. 



If I utilize the new `samples: 12` I get the following result with a width of <=0.04°C. 



Also, it seems the data scattering is not entirely caused by white noise and the `samples` improved the result more than the expected 1/sqrt(N), when modifying N even changing bucket size and measurement resolution to adjust for the narrower Gaussian. Consequently, I think this feature provides a qualitative improvement compared to, e.g., an averaging filter.

File: esphome/components/adc/sensor.py [hunk 5]
```diff
@@ -67,6 +69,7 @@ def final_validate_config(config):
             cv.SplitDefault(CONF_ATTENUATION, esp32="0db"): cv.All(
                 cv.only_on_esp32, cv.enum(ATTENUATION_MODES, lower=True)
             ),
+            cv.Optional(CONF_SAMPLES, default=1): cv.int_range(min=1, max=255),
```
Comment (jesserockz) [inline esphome/components/adc/sensor.py:90]: Please add a custom validation that checks the attenuation on esp32 and raises an error if the value is not 1

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Comment by edwardtfn** [comment]
I'm getting this when compiling with ESPHome 2024.5.0b4:

```log

src/esphome/components/adc/adc_sensor.cpp: In member function 'virtual void esphome::adc::ADCSensor::setup()':
src/esphome/components/adc/adc_sensor.cpp:65:28: warning: 'ADC_ATTEN_DB_11' is deprecated [-Wdeprecated-declarations]
   for (int32_t i = 0; i <= ADC_ATTEN_DB_11; i++) {
                            ^~~~~~~~~~~~~~~
In file included from /data/cache/platformio/packages/framework-espidf@3.40407.0/components/driver/include/driver/adc.h:14,
                 from src/esphome/components/adc/adc_sensor.h:10,
                 from src/esphome/components/adc/adc_sensor.cpp:1:
/data/cache/platformio/packages/framework-espidf@3.40407.0/components/hal/include/hal/adc_types.h:54:5: note: declared here
     ADC_ATTEN_DB_11 __attribute__((deprecated)) = ADC_ATTEN_DB_12,  ///<This is deprecated, it behaves the same as `ADC_ATTEN_DB_12`
     ^~~~~~~~~~~~~~~
src/esphome/components/adc/adc_sensor.cpp: In member function 'virtual void esphome::adc::ADCSensor::dump_config()':
src/esphome/components/adc/adc_sensor.cpp:121:12: warning: 'ADC_ATTEN_DB_11' is deprecated [-Wdeprecated-declarations]
       case ADC_ATTEN_DB_11:
            ^~~~~~~~~~~~~~~
In file included from /data/cache/platformio/packages/framework-espidf@3.40407.0/components/driver/include/driver/adc.h:14,
                 from src/esphome/components/adc/adc_sensor.h:10,
                 from src/esphome/components/adc/adc_sensor.cpp:1:
/data/cache/platformio/packages/framework-espidf@3.40407.0/components/hal/include/hal/adc_types.h:54:5: note: declared here
     ADC_ATTEN_DB_11 __attribute__((deprecated)) = ADC_ATTEN_DB_12,  ///<This is deprecated, it behaves the same as `ADC_ATTEN_DB_12`
     ^~~~~~~~~~~~~~~
src/esphome/components/adc/adc_sensor.cpp: In member function 'virtual float esphome::adc::ADCSensor::sample()':
src/esphome/components/adc/adc_sensor.cpp:203:48: warning: 'ADC_ATTEN_DB_11' is deprecated [-Wdeprecated-declarations]
     adc1_config_channel_atten(this->channel1_, ADC_ATTEN_DB_11);
                                                ^~~~~~~~~~~~~~~
In file included from /data/cache/platformio/packages/framework-espidf@3.40407.0/components/driver/include/driver/adc.h:14,
                 from src/esphome/components/adc/adc_sensor.h:10,
                 from src/esphome/components/adc/adc_sensor.cpp:1:
/data/cache/platformio/packages/framework-espidf@3.40407.0/components/hal/include/hal/adc_types.h:54:5: note: declared here
     ADC_ATTEN_DB_11 __attribute__((deprecated)) = ADC_ATTEN_DB_12,  ///<This is deprecated, it behaves the same as `ADC_ATTEN_DB_12`
     ^~~~~~~~~~~~~~~
src/esphome/components/adc/adc_sensor.cpp:218:48: warning: 'ADC_ATTEN_DB_11' is deprecated [-Wdeprecated-declarations]
     adc2_config_channel_atten(this->channel2_, ADC_ATTEN_DB_11);
                                                ^~~~~~~~~~~~~~~
In file included from /data/cache/platformio/packages/framework-espidf@3.40407.0/components/driver/include/driver/adc.h:14,
                 from src/esphome/components/adc/adc_sensor.h:10,
                 from src/esphome/components/adc/adc_sensor.cpp:1:
/data/cache/platformio/packages/framework-espidf@3.40407.0/components/hal/include/hal/adc_types.h:54:5: note: declared here
     ADC_ATTEN_DB_11 __attribute__((deprecated)) = ADC_ATTEN_DB_12,  ///<This is deprecated, it behaves the same as `ADC_ATTEN_DB_12`
     ^~~~~~~~~~~~~~~
src/esphome/components/adc/adc_sensor.cpp:238:91: warning: 'ADC_ATTEN_DB_11' is deprecated [-Wdeprecated-declarations]
   uint32_t mv11 = esp_adc_cal_raw_to_voltage(raw11, &this->cal_characteristics_[(int32_t) ADC_ATTEN_DB_11]);
                                                                                           ^~~~~~~~~~~~~~~
In file included from /data/cache/platformio/packages/framework-espidf@3.40407.0/components/driver/include/driver/adc.h:14,
                 from src/esphome/components/adc/adc_sensor.h:10,
                 from src/esphome/components/adc/adc_sensor.cpp:1:
/data/cache/platformio/packages/framework-espidf@3.40407.0/components/hal/include/hal/adc_types.h:54:5: note: declared here
     ADC_ATTEN_DB_11 __attribute__((deprecated)) = ADC_ATTEN_DB_12,  ///<This is deprecated, it behaves the same as `ADC_ATTEN_DB_12`
     ^~~~~~~~~~~~~~~
```

**Comment by jesserockz** [comment]
@edwardtfn 
Unrelated to this PR. The field was changed in esp-idf 4.4.7
https://github.com/espressif/esp-idf/commit/9b2661c91cbfa5dfba8da0311fdd9bf204fd1730
Needs to be fixed in its own PR in ESPHome.
Can you please report a new issue for it?

**Review by jesserockz — APPROVED** [review]

**Pull request merged**